packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -824,6 +834,24 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             message.contents = JSON.parse(message.contents);
         }
 
+        // validate client sequence number has no gap. If there is gap, check if there were noops
+        // if there were noops, decrement the gap by number of noops and continue
+        if (this.connectionManager.lastSubmittedClientId !== undefined && this.connectionManager.lastSubmittedClientId === message.clientId) {
+            assert(this.lastClientSequenceNumber !== undefined, "lastClientSequenceNumber should not be undefined");
+            if (message.type === MessageType.NoOp){
+                this.noOpCount--;
+            }
+            const clientSeqNumGap = message.clientSequenceNumber - this.lastClientSequenceNumber - 1;
+            if (clientSeqNumGap > 0) {
+                if (this.noOpCount > 0) {
+                    this.noOpCount -= clientSeqNumGap;
+                } else {
+                    throw new Error(`gap in client sequence number :${clientSeqNumGap}`);
+                }
+            }
+        }
+        this.lastClientSequenceNumber = message.clientSequenceNumber;
```

Comment:
I believe you can bump this.lastClientSequenceNumber only inside of IF block above, i.e. only if message is from this client. Otherwise, if message is for different client, this will be rather random number, that will screw up the rest of the logic. This should fail tests, as it would be rather easy to hit these conditions where next ops (from this client) would result in some asserts / incorrect state.